src/VisualStudio/Core/Def/Telemetry/AggregatingTelemetryLog.cs

```diff
@@ -0,0 +1,117 @@
+
+        private const string MeterVersion = "0.39";
+
+        public AggregatingTelemetryLog(TelemetrySession session, FunctionId functionId, double[]? bucketBoundaries)
+        {
+            var meterName = TelemetryLogger.GetPropertyName(functionId, "meter");
+            var meterProvider = new VSTelemetryMeterProvider();
+
+            _session = session;
+            _meter = meterProvider.CreateMeter(meterName, version: MeterVersion);
+            _histograms = new();
+            _lock = new();
+            _eventName = TelemetryLogger.GetEventName(functionId);
+
+            if (bucketBoundaries != null)
+            {
+                _histogramConfiguration = new HistogramConfiguration(bucketBoundaries);
+            }
+        }
+
+        /// <summary>
+        /// Adds aggregated information for the metric and value passed in via logMessage. The Name/Value properties
+        /// are used as the metric name and value to record.
+        /// </summary>
+        /// <param name="logMessage"></param>
+        public void Log(LogMessage logMessage)
+        {
+            const string Name = nameof(Name);
+            const string Value = nameof(Value);
+
+            if (!IsEnabled)
+                return;
+
+            if (logMessage is not KeyValueLogMessage kvLogMessage)
+                return;
+
+            if (!kvLogMessage.TryGetValue(Name, out var nameValue) || nameValue is not string metricName)
+                throw ExceptionUtilities.Unreachable();
+
+            if (!kvLogMessage.TryGetValue(Value, out var valueValue) || valueValue is not int value)
+                throw ExceptionUtilities.Unreachable();
+
+            IHistogram<int>? histogram;
+            lock (_lock)
+            {
+                if (!_histograms.TryGetValue(metricName, out histogram))
+                {
+                    histogram = _meter.CreateHistogram<int>(metricName, _histogramConfiguration);
+                    _histograms.Add(metricName, histogram);
+                }
+            }
+
+            histogram.Record(value);
+        }
+
+        public IDisposable? LogBlockTime(string name, int minThreshold)
+        {
+            if (!IsEnabled)
+                return null;
+
+            return new TimedTelemetryLogBlock(name, minThreshold, this);
+        }
+
+        private bool IsEnabled
+        {
+            get
+            {
+                return _session.IsOptedIn;
+            }
+        }
+
+        public void PostTelemetry(TelemetrySession session)
+        {
+            lock (_lock)
+            {
+                foreach (var histogram in _histograms.Values)
+                {
+                    var telemetryEvent = new TelemetryEvent(_eventName);
+                    var histogramEvent = new TelemetryHistogramEvent<int>(telemetryEvent, histogram);
+
+                    session.PostMetricEvent(histogramEvent);
+                }
+
+                _histograms.Clear();
```

Comment:
i would strongly consider making a copy of hte data within the lock (and clearing), then posting *outside* of the lock.

i do not like calling to other services within a lock here.

Reply:
Yep, good catch. Commit 7.